### PR TITLE
reduce http2 complexity

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Reducing instructions needed for http2 parsing program.

### Motivation
As part of the effort to allow running http2 decoding on older kernels than 5.2, we are introducing the first phase, in which we simplify the program, and reduce instructions count by 62% in total.
We achieve that by re-ordering functions and removing redundant calls.
The changes are divided into self-contained commits with explanation and the reduction size for each change.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/protocols/http/model_http2_linux.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ func (tx *ebpfHttp2Tx) RequestLatency() float64 {
 // Incomplete returns true if the transaction contains only the request or response information
 // This happens in the context of localhost with NAT, in which case we join the two parts in userspace
 func (tx *ebpfHttp2Tx) Incomplete() bool {
-	return tx.Request_started == 0 || tx.Response_status_code == 0 || len(tx.Request_path) == 0
+	return tx.Request_started == 0 || tx.Response_status_code == 0 || tx.Path_size == 0 || (tx.Request_method != GetValue && tx.Request_method != PostValue)
```
Comment (amitslavin) [inline pkg/network/protocols/http/model_http2_linux.go:44]: Do we want to add this functionality for this flow? HTTP/2 can be more then POST or GET

**Review by amitslavin — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/http/model_http2_linux.go (hunk 1 above)
Comment (guyarb) [reply pkg/network/protocols/http/model_http2_linux.go:44]: changed

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 2]
```diff
@@ -39,284 +50,232 @@ static __always_inline http2_stream_t *http2_fetch_stream(http2_stream_key_t *ht
...
-    str_len = 0;
-    if (!read_var_int(heap_buffer, 6, &str_len, stream_id)) {
-        return HEADER_ERROR;
-    }
-
-    // if the index is not path or the len of string is bigger then we support, we continue.
-    if (str_len >= HTTP2_MAX_PATH_LEN || index != kIndexPath){
-        heap_buffer->offset += str_len;
-        return HEADER_NOT_INTERESTING;
+    if (str_len >= HTTP2_MAX_PATH_LEN || index != kIndexPath || headers_to_process == NULL){
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:145]: changed `>=` into `>`

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 3]
```diff
@@ -39,284 +50,232 @@ static __always_inline http2_stream_t *http2_fetch_stream(http2_stream_key_t *ht
...
 
+static __always_inline bool read_var_int(struct __sk_buff *skb, skb_info_t *skb_info, __u8 max_number_for_bits, __u8 *out){
+    if (skb_info->data_off > skb->len) {
+        return false;
+    }
+    __u8 current_char_as_number = 0;
+    bpf_skb_load_bytes(skb, skb_info->data_off, &current_char_as_number, sizeof(current_char_as_number));
+    skb_info->data_off++;
+
+    return read_var_int_2(skb, skb_info, current_char_as_number, max_number_for_bits, out);
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:81]: renamed and added doc

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 4]
```diff
@@ -39,284 +50,232 @@ static __always_inline http2_stream_t *http2_fetch_stream(http2_stream_key_t *ht
...
-        return false;
-    }
-    // TODO: verifier is happy now.
-    if (HTTP2_BUFFER_SIZE-1 <= offset) {
-        return false;
-    }
-    const __u8 current_char_as_number_s = heap_buffer->fragment[offset];
-    __u8 current_char_as_number = current_char_as_number_s;
-    current_char_as_number &= (1 << factor) - 1;
+static __always_inline bool read_var_int_2(struct __sk_buff *skb, skb_info_t *skb_info, __u8 current_char_as_number, __u8 max_number_for_bits, __u8 *out){
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:52]: done

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 5]
```diff
@@ -39,284 +50,232 @@ static __always_inline http2_stream_t *http2_fetch_stream(http2_stream_key_t *ht
...
         *out = current_char_as_number;
         return true;
     }
 
-    const u16 offset2 = heap_buffer->offset;
-    if (offset2 < heap_buffer->size && offset2 < HTTP2_BUFFER_SIZE) {
-        const __u8 b = heap_buffer->fragment[offset2];
-        current_char_as_number += b & 127;
+    if (skb_info->data_off <= skb->len) {
+        __u8 b = 0;
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:61]: done

**Review by amitslavin — APPROVED** [review]

**Pull request merged**